### PR TITLE
soc_bmwbc: add new captcha for initial login

### PR DESCRIPTION
**Pull request by rleidner**
Support new captcha required by BMW server login
Store token file in folder data/bmwbc instead of ramdisk to avoid too frequent captcha process after reboot.
related UI PR: https://github.com/openWB/openwb-ui-settings/pull/589

### CODE REVIEW

File: packages/modules/vehicles/bmwbc/api.py [hunk 1]
```diff
@@ -20,6 +20,9 @@
 
 from modules.common.component_state import CarState
 from modules.common.store import RAMDISK_PATH
+from pathlib import Path
+
+dataPath = "/var/www/html/openWB/data/bmwbc"
```
Comment (benderl) [inline packages/modules/vehicles/bmwbc/api.py:25]: Bitte keinen absoluten Pfad verwenden. Analog zu `RAMDISK_PATH` sollte es auch eine Konstante für das Wurzelverzeichnis der Installation geben, habe ich aber gerade nicht im Kopf. Alternativ das Verzeichnis relativ zur aktuellen Datei ermitteln.

**Review by benderl — CHANGES_REQUESTED** [review]

File: packages/modules/vehicles/bmwbc/api.py (hunk 1 above)
Comment (rleidner) [reply packages/modules/vehicles/bmwbc/api.py]: Ist erledigt. Die Konstante habe ich auch nicht gefunden, ist jetzt relativ.

File: packages/modules/vehicles/bmwbc/api.py [hunk 2]
```diff
@@ -22,7 +22,7 @@
 from modules.common.store import RAMDISK_PATH
 from pathlib import Path
 
-dataPath = "/var/www/html/openWB/data/bmwbc"
+DATA_PATH = "data/bmwbc"
```
Comment (benderl) [inline packages/modules/vehicles/bmwbc/api.py:25]: Es ist einfacher, wenn Du mit dem Path-Objekt arbeitest.
```suggestion
DATA_PATH = Path(__file__).resolve().parents[4] / "data" / "bmwbc"
```

File: packages/modules/vehicles/bmwbc/api.py [hunk 3]
```diff
@@ -80,6 +73,16 @@ def dump_json(data: dict, fout: str):
...
 async def _fetch_soc(user_id: str, password: str, vin: str, captcha_token: str, vnum: int) -> Union[int, float]:
     global storeFile
+    try:
+        openwbpath = str(Path(__file__).resolve().parents[4])
+        dataPath = openwbpath + '/' + DATA_PATH
+        log.debug("dataPath=" + dataPath)
+        Path(dataPath).mkdir(parents=True, exist_ok=True)
+    except Exception as e:
+        log.error("init: dataPath creation failed, dataPath: " +
+                  dataPath + ", error=" + str(e))
```
Comment (benderl) [inline packages/modules/vehicles/bmwbc/api.py:83]:
```suggestion
        log.debug("dataPath=" + str(DATA_PATH))
        DATA_PATH.mkdir(parents=True, exist_ok=True)
    except Exception as e:
        log.error("init: dataPath creation failed, dataPath: " +
                  str(DATA_PATH) + ", error=" + str(e))
```

**Review by benderl — CHANGES_REQUESTED** [review]

File: packages/modules/vehicles/bmwbc/api.py (hunk 3 above)
Comment (rleidner) [reply packages/modules/vehicles/bmwbc/api.py]: Ich habe es bei mir so geändert, getestet und ein neues commit gemacht.

File: packages/modules/vehicles/bmwbc/api.py (hunk 2 above)
Comment (rleidner) [reply packages/modules/vehicles/bmwbc/api.py]: Ich habe es bei mir so geändert, getestet und ein neues commit gemacht.

**Review by benderl — APPROVED** [review]

**Comment by Meli73** [comment]
In OWB1 nightly It works fine with Captcha.
Is there already a way to test it here in OWB2?   in Alpha, Beta or Release ?

**Comment by Meli73** [comment]
tried it , but got this:

2024-12-12 14:53:19,334 - {modules.vehicles.bmwbc.api:49} - {WARNING:fetch soc_ev6} - load_store: store file not found, full authentication required
2024-12-12 14:53:19,335 - {modules.vehicles.bmwbc.api:143} - {ERROR:fetch soc_ev6} - bmwbc.fetch_soc: requestData Error, vnum: 6 err=TypeError("__init__() got an unexpected keyword argument 'hcaptcha_token'"), type(err)=<class 'TypeError'>
2024-12-12 14:53:19,369 - {modules.common.fault_state:49} - {ERROR:fetch soc_ev6} - BMW (Bimmer): FaultState FaultStateLevel.ERROR, FaultStr <class 'TypeError'> ("__init__() got an unexpected keyword argument 'hcaptcha_token'",), Traceback: 
Traceback (most recent call last):
  File "/var/www/html/openWB/packages/modules/common/configurable_vehicle.py", line 66, in update
    car_state = self._get_carstate_by_source(vehicle_update_data, source)
  File "/var/www/html/openWB/packages/modules/common/configurable_vehicle.py", line 110, in _get_carstate_by_source
    return self.__component_updater(vehicle_update_data)
  File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/soc.py", line 20, in updater
    return api.fetch_soc(
  File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/api.py", line 156, in fetch_soc
    soc, range = loop.run_until_complete(_fetch_soc(user_id, password, vin, captcha_token, vnum))
  File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete
    return future.result()
  File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/api.py", line 101, in _fetch_soc
    auth = MyBMWAuthentication(user_id,
TypeError: __init__() got an unexpected keyword argument 'hcaptcha_token'

**Comment by rleidner** [comment]
> tried it , updated to 2.1.7-Alpha.1 2024-12-12 15:49:02 +0100 [[1927094](https://github.com/openWB/core/commit/192709443f5040ba102c1377b36e2ad50d7c2ff3)]
> 
> but got this:
> 
> 2024-12-12 14:53:19,334 - {modules.vehicles.bmwbc.api:49} - {WARNING:fetch soc_ev6} - load_store: store file not found, full authentication required 2024-12-12 14:53:19,335 - {modules.vehicles.bmwbc.api:143} - {ERROR:fetch soc_ev6} - bmwbc.fetch_soc: requestData Error, vnum: 6 err=TypeError("**init**() got an unexpected keyword argument 'hcaptcha_token'"), type(err)=<class 'TypeError'> 2024-12-12 14:53:19,369 - {modules.common.fault_state:49} - {ERROR:fetch soc_ev6} - BMW (Bimmer): FaultState FaultStateLevel.ERROR, FaultStr <class 'TypeError'> ("**init**() got an unexpected keyword argument 'hcaptcha_token'",), Traceback: Traceback (most recent call last): File "/var/www/html/openWB/packages/modules/common/configurable_vehicle.py", line 66, in update car_state = self._get_carstate_by_source(vehicle_update_data, source) File "/var/www/html/openWB/packages/modules/common/configurable_vehicle.py", line 110, in _get_carstate_by_source return self.__component_updater(vehicle_update_data) File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/soc.py", line 20, in updater return api.fetch_soc( File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/api.py", line 156, in fetch_soc soc, range = loop.run_until_complete(_fetch_soc(user_id, password, vin, captcha_token, vnum)) File "/usr/lib/python3.9/asyncio/base_events.py", line 642, in run_until_complete return future.result() File "/var/www/html/openWB/packages/modules/vehicles/bmwbc/api.py", line 101, in _fetch_soc auth = MyBMWAuthentication(user_id, TypeError: **init**() got an unexpected keyword argument 'hcaptcha_token'

The PR is now merged - please update to  latest master.

**Comment by Meli73** [comment]
you mean :
updated to 2.1.7-Alpha.1 2024-12-12 15:49:02 +0100 [[1927094](https://github.com/openWB/core/commit/192709443f5040ba102c1377b36e2ad50d7c2ff3)]

I did already..  or is there somthing newer  ?

**Comment by rleidner** [comment]
> you mean : updated to 2.1.7-Alpha.1 2024-12-12 15:49:02 +0100 [[1927094](https://github.com/openWB/core/commit/192709443f5040ba102c1377b36e2ad50d7c2ff3)]
> 
> I did already.. or is there somthing newer ?

That version should be working; if you continue to have issues, you get support in the openwb forum. 
I can't maintain support on 2 different platforms; therefore I will not respond here from now on.

**Pull request merged**